### DISCUSSION TITLE
using dynamic dispatch to implement sui-style Programmable Transaction Blocks (PTBs)

### DISCUSSION THREAD

**Discussion by MoonShiesty** [aptos-labs/aptos-developer-discussions#281] [General]
now that dynamic dispatch is live, we can implement sui-style PTBs in the aptos runtime using dynamic dispatch. while similar functionality can be achieved using move scripts, PTBs can be constructed at runtime without invoking the move compiler. this feature is heavily used in the SUI ecosystem to construct complex, multi-call transactions such as swap aggregation, on-chain arbitrage or aggregating FAs transfer from multiple stores (this is a growing problem in the aptos ecosystem, especially with the coin->fa migration).

what is a sui PTB:

> is comprised of individual transaction commands (sometimes referred to simply as transactions or commands). Each transaction command executes in order, and you can use the results from one transaction command in any subsequent transaction command. The effects, specifically object modifications or transfers, of all transaction commands in a block are applied atomically at the end of the transaction. If one transaction command fails, the entire block fails and no effects from the commands are applied.

why should aptos implement PTBs:

the aptos roadmap calls for using dynamic dispatch in specific cases where it has benefits to the developer without adding unacceptable security risk. since PTBs can only be used at the entrypoint of a move call, they introduce a far smaller security footprint than generalized dynamic dispatch, while improving the developer experience.

downsides:

the downside to PTBs is that they aren't checked by the move compiler. this adds security risk relative to move scripts, but i think we'll always face this trade-off when adding more ways to use dynamic dispatch

im in the process of writing an AIP for aptos PTBs. i wanted to solicit some feedback here first, curious what you all think!

**Comment by gregnazario** [comment]
There's already an AIP in progress here @MoonShiesty https://github.com/aptos-foundation/AIPs/pull/448

**Reply by MoonShiesty** [reply]
just took a look at the AIP, awesome stuff
